go/glome/glome.go

```diff
@@ -0,0 +1,191 @@
+type PublicKey [PublicKeySize]byte
+
+// Generate a PublicKey object from slice. Return ErrInvalidPublicKey
+// if slice's length is not PublicKeySize.
+func PublicKeyFromSlice(b []byte) (*PublicKey, error) {
+	if len(b) != PublicKeySize {
+		return nil, ErrInvalidPublicKey
+	}
+
+	var p PublicKey
+	copy(p[:], b)
+	return &p, nil
+}
+
+// PrivateKey is the type of GLOME private keys.
+type PrivateKey [PrivateKeySize]byte
+
+// Public returns the PublicKey corresponding to priv.
+func (priv *PrivateKey) Public() (*PublicKey, error) {
+	slice, err := curve25519.X25519(priv[:], curve25519.Basepoint)
+	if err != nil {
+		return nil, err
+	}
+
+	p, _ := PublicKeyFromSlice(slice)
+	return p, nil
+}
+
+// Generates a Dialog struct. It performs GLOME handshake, and stores create a Dialog
+// from the user to the peer. Sets MinPeerTagSize as MaxTagSize.
+func (user *PrivateKey) Exchange(peer *PublicKey) (*Dialog, error) {
+	s, err := curve25519.X25519(user[:], peer[:])
+	if err != nil {
+		return nil, err
+	}
+
+	public, err := user.Public()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Dialog{shared: s, User: *public, Peer: *peer, MinPeerTagSize: MaxTagSize}, nil
+}
+
+// Generates a Dialog struct. It performs GLOME handshake, and stores create a Dialog
+// from the user to the peer. Sets param m as MinPeerTagSize.
+func (user *PrivateKey) TruncatedExchange(peer *PublicKey, m uint) (*Dialog, error) {
+	if m == 0 || m > MaxTagSize {
+		return nil, ErrInvalidTagSize
+	}
+
+	d, err := user.Exchange(peer)
+	if err != nil {
+		return nil, err
+	}
+
+	d.MinPeerTagSize = m
+	return d, nil
+}
+
+// Generates a private key from a slice. Fail if len of slice is not PrivateKeySize
+func PrivateKeyFromSlice(b []byte) (*PrivateKey, error) {
+	if len(b) != PrivateKeySize {
+		return nil, ErrInvalidPrivateKey
+	}
+
+	var p PrivateKey
+	copy(p[:], b)
+	return &p, nil
+}
+
+// GenerateKey generates a public/private key pair using entropy from rand.
+func GenerateKeys(rand io.Reader) (*PublicKey, *PrivateKey, error) {
+	b := make([]byte, PrivateKeySize)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	priv, err := PrivateKeyFromSlice(b)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pub, err := priv.Public()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pub, priv, nil
+}
+
+// Implement tag managing functionalities for GLOME protocol.
+type Dialog struct {
+	shared         []byte
+	User           PublicKey // User's Public key
+	Peer           PublicKey // Peer's Public key
+	MinPeerTagSize uint      // Minimun Tag size allowed.
+}
+
+func (d *Dialog) sendingKey() []byte {
+	return append(d.shared[:], append(d.Peer[:], d.User[:]...)...)
+}
+
+func (d *Dialog) receivingKey() []byte {
+	return append(d.shared[:], append(d.User[:], d.Peer[:]...)...)
+}
+
+// Generates a tag matching some provided message, counter and password.
+func generate_tag(msg []byte, counter uint8, password []byte) []byte {
```

Comment:
shouldn't it be `generateKey`?

Reply:
Yes, it has an error. It should be generateTag.

Reply:
Fixed.